src/crypto/crypto.js

```diff
@@ -152,83 +151,141 @@ export default {
-      //      - MPI of DSA secret exponent x.
-      case enums.publicKey.dsa:
-        return [type_mpi];
-      //   Algorithm-Specific Fields for ECDSA or ECDH secret keys:
-      //       - MPI of an integer representing the secret key.
-      case enums.publicKey.ecdh:
-      case enums.publicKey.ecdsa:
-      case enums.publicKey.eddsa:
-        return [type_mpi];
-      default:
-        throw new Error('Invalid public key encryption algorithm.');
-    }
-  },
-
-  /** Returns the types comprising the public key of an algorithm
-   * @param {module:enums.publicKey}  algo The public key algorithm
-   * @returns {Array<Object>}         The array of types
+  /**
+   * Parse public key material in binary form to get the key parameters
+   * @param {module:enums.publicKey} algo  The key algorithm
+   * @param {Uint8Array}             bytes The key material to parse
+   * @returns {Object} key parameters referenced by name
+   * @returns { read: Number, params: Object } number of read bytes plus key parameters referenced by name
    */
-  getPubKeyParamTypes: function(algo) {
+  parsePublicKeyParams: function(algo, bytes) {
+    let read = 0;
     switch (algo) {
       //   Algorithm-Specific Fields for RSA public keys:
       //       - a multiprecision integer (MPI) of RSA public modulus n;
       //       - an MPI of RSA public encryption exponent e.
-      case enums.publicKey.rsaEncrypt:
-      case enums.publicKey.rsaEncryptSign:
-      case enums.publicKey.rsaSign:
-        return [type_mpi, type_mpi];
-      //   Algorithm-Specific Fields for Elgamal public keys:
-      //       - MPI of Elgamal prime p;
-      //       - MPI of Elgamal group generator g;
-      //       - MPI of Elgamal public key value y (= g**x mod p where x  is secret).
-      case enums.publicKey.elgamal:
-        return [type_mpi, type_mpi, type_mpi];
       //   Algorithm-Specific Fields for DSA public keys:
       //       - MPI of DSA prime p;
       //       - MPI of DSA group order q (q is a prime divisor of p-1);
       //       - MPI of DSA group generator g;
       //       - MPI of DSA public-key value y (= g**x mod p where x  is secret).
+      //   Algorithm-Specific Fields for Elgamal public keys:
+      //       - MPI of Elgamal prime p;
+      //       - MPI of Elgamal group generator g;
+      //       - MPI of Elgamal public key value y (= g**x mod p where x  is secret).
+      case enums.publicKey.rsaEncrypt:
+      case enums.publicKey.rsaEncryptSign:
+      case enums.publicKey.rsaSign:
       case enums.publicKey.dsa:
-        return [type_mpi, type_mpi, type_mpi, type_mpi];
+      case enums.publicKey.elgamal: {
+        const params = {};
+        const paramNames = this.getPublicKeyParamNames(algo);
+        paramNames.forEach(name => {
+          const param = util.readMPI(bytes.subarray(read)); read += param.length + 2;
+          params[name] = param;
+        });
+        return { read: read, params };
+      }
```

Comment:
And then here, I would split up the cases and parse each param explicitly; I know it's more repetitive but then it's more consistent with the code below, and once we switch to streaming (and can have a `parseMPI` function that consumes the stream), it won't be too bad.
(And similarly for the private params.)